resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -559,4 +739,118 @@ private void addTrace(Throwable cause) {
+
+            int dots = 0;
+            for (int a = recordName.length() - 1, b = questionName.length() - 1; a >= 0; a--, b--) {
+                char c = recordName.charAt(a);
+                if (questionName.charAt(b) != c) {
+                    return;
+                }
+                if (c == '.') {
+                    dots++;
+                }
+            }
+
+            if (head != null && head.dots > dots) {
+                // We already have a closer match so ignore this one, no need to parse the domainName etc.
+                return;
+            }
+
+            final ByteBuf recordContent = ((ByteBufHolder) r).content();
+            final String domainName = decodeDomainName(recordContent);
+            if (domainName == null) {
+                // Could not be parsed, ignore.
+                return;
+            }
+
+            // We are only interested in preserving the nameservers which are the closest to our qName, so ensure
+            // we drop servers that have a smaller dots count.
+            if (head == null || head.dots < dots) {
+                count = 1;
+                head = new AuthoritativeNameServer(dots, recordName, domainName);
+            } else if (head.dots == dots) {
+                AuthoritativeNameServer serverName = head;
+                while (serverName.next != null) {
+                    serverName = serverName.next;
+                }
+                serverName.next = new AuthoritativeNameServer(dots, recordName, domainName);
+                count++;
+            }
+        }
+
+        // Just walk the linked-list and mark the entry as removed when matched, so next lookup will need to process
+        // one node less.
+        AuthoritativeNameServer remove(String nsName) {
+            AuthoritativeNameServer serverName = head;
+
+            while (serverName != null) {
+                if (!serverName.removed && serverName.nsName.equalsIgnoreCase(nsName)) {
+                    serverName.removed = true;
+                    return serverName;
+                }
+                serverName = serverName.next;
+            }
+            return null;
+        }
+
+        int size() {
+            return count;
+        }
+    }
+
+    static final class AuthoritativeNameServer {
+        final int dots;
+        final String nsName;
+        final String domainName;
+
+        AuthoritativeNameServer next;
+        boolean removed;
+
+        AuthoritativeNameServer(int dots, String domainName, String nsName) {
+            this.dots = dots;
+            this.nsName = nsName;
+            this.domainName = domainName;
+        }
+
+        /**
+         * Returns {@code true} if its a root server.
+         */
+        boolean isRootServer() {
+            return dots == 1;
+        }
+
+        /**
+         * The domain for which the {@link AuthoritativeNameServer} is responsible.
+         */
+        String domainName() {
+            return domainName;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Rename method "domainName" to prevent any misunderstanding/clash with field "domainName" defined on line 832. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1845)